lib/ramble/ramble/application.py

```diff
@@ -650,13 +656,84 @@ def _get_executables(self):
             for name, conf in self.internals[namespace.custom_executables].items():
                 self.executables[name] = ramble.util.executable.CommandExecutable(
                     name=name,
                     **conf
                 )
 
+        # Perform executable injection
+        if namespace.executable_injection in self.internals:
+
+            supported_orders = set(['before', 'after'])
+
+            # Order can be 'before' or 'after.
+            # If `relative_to` is not set, then before adds to be the beginning of the list
+            #                  and after (default) adds to the end of the list
+            # If `relative_to` IS set, then before adds before the first instance of
+            #                  the executable in the list
+            #                  and after (default) adds after the last instance of the
+            #                  executable in the list
+            # If `relative_to` is set, and the executable name is not found, raise a fatal error.
+            for exec_injection in self.internals[namespace.executable_injection]:
+                exec_name = exec_injection['name']
+
+                injection_order = 'after'
+                if 'order' in exec_injection:
+                    injection_order = exec_injection['order']
+
+                relative = None
+                if 'relative_to' in exec_injection:
+                    relative = exec_injection['relative_to']
+
+                if injection_order not in supported_orders:
+                    logger.die('In experiment '
+                               f'"{self.expander.experiment_namespace}" '
+                               f'injection order of executable "{exec_name}" is set to an '
+                               f'invalid value of "{injection_order}".\n'
+                               f'Valid values are {supported_orders}.')
+
+                if exec_name not in self.executables:
+                    logger.die('In experiment '
+                               f'"{self.expander.experiment_namespace}" '
+                               f'attempting to inject a non existing executable "{exec_name}".')
+
+                if relative is not None and relative not in executables:
+                    logger.die('In experiment '
+                               f'"{self.expander.experiment_namespace}" '
+                               f'attempting to inject executable "{exec_name}" '
+                               f'relative to a non existing executable "{relative}".')
+
+                if relative is None:
+                    if injection_order == 'before':
+                        executables.insert(0, exec_name)
+                    elif injection_order == 'after':
+                        executables.append(exec_name)
+                else:
+
+                    found = False
+                    if injection_order == 'before':
+                        relative_index = 0
```

Comment:
nit: all this relative index rejiggering doesn't feel very pythonic.. but I am struggling to come up with a definitely better solution

I'm kind of surprised you did not turn this into a more graph-style solution where you walk it and just append the pre/node/post and went right to the 1d splat of that.. unless you have some awesome solution lets leave this as-is for now?

Reply:
There are lots of places a graph would be nice for. Maybe we should add that to the task list and use it to clean things up in the future.